File: labs-opensource/M15-code-execution-sandbox/solution/sandbox_executor.py
"""
M15 Lab - Step 1: SubprocessExecutor — SOLUTION
================================================
Run: python sandbox_executor.py
"""

import os
import subprocess
import sys
import tempfile
from dataclasses import dataclass


@dataclass
class ExecResult:
    stdout: str
    stderr: str
    exit_code: int

    def to_tool_content(self) -> str:
        """Format the result as a concise tool message for the model."""
        lines = [f"exit_code: {self.exit_code}"]
        if self.stdout:
            lines.append(f"stdout:\n{self.stdout.strip()}")
        if self.stderr:
            lines.append(f"stderr:\n{self.stderr.strip()}")
        return "\n".join(lines)


class SubprocessExecutor:
    """Runs Python code in a child process.

    Dev-tier sandbox: zero Docker overhead, but FULL filesystem access.
    Never use for untrusted user input — upgrade to DockerExecutor.
    """

    def __init__(self, timeout_seconds: int = 10):
        self.timeout_seconds = timeout_seconds

    def run(self, code: str, timeout: int | None = None) -> ExecResult:
        """Execute code, return ExecResult. Never raises."""
        timeout = timeout or self.timeout_seconds

        # Temp file gives the child its own scope; delete=False so the
        # child can open it after we close it.
        with tempfile.NamedTemporaryFile(
            mode="w", suffix=".py", delete=False, encoding="utf-8"
        ) as f:
            f.write(code)
            tmp_path = f.name

        try:
            result = subprocess.run(
                [sys.executable, tmp_path],
                capture_output=True,
                text=True,
                timeout=timeout,
                shell=False,  # CRITICAL: never True (shell injection)
            )
            return ExecResult(
                stdout=result.stdout,
                stderr=result.stderr,
                exit_code=result.returncode,
            )
        except subprocess.TimeoutExpired:
            return ExecResult(
                stdout="",
                stderr=f"Execution timed out after {timeout} seconds.",
                exit_code=124,  # bash timeout(1) convention
            )
        except Exception as e:
            return ExecResult(stdout="", stderr=str(e), exit_code=1)
        finally:
            os.unlink(tmp_path)  # cleanup even on crash


if __name__ == "__main__":
    ex = SubprocessExecutor(timeout_seconds=3)

    print("TEST 1: happy path")
    r = ex.run("print(sum(range(101)))")
    print(f"  {r.to_tool_content()}")
    assert r.exit_code == 0 and "5050" in r.stdout

    print("\nTEST 2: deliberate NameError (must return, not raise)")
    r = ex.run("print(undefined_variable)")
    print(f"  exit_code={r.exit_code}, stderr contains NameError: {'NameError' in r.stderr}")
    assert r.exit_code != 0 and "NameError" in r.stderr

    print("\nTEST 3: infinite loop (must time out at exit code 124, not hang)")
    r = ex.run("while True: pass")
    print(f"  exit_code={r.exit_code}, stderr={r.stderr[:50]}")
    assert r.exit_code == 124

    print("\nAll executor checks passed.")

File: labs-opensource/M15-code-execution-sandbox/solution/sandbox_executor.js
/**
 * M15 Lab - Step 1: SubprocessExecutor — SOLUTION (Node.js)
 * ==========================================================
 * Run: node sandbox_executor.js
 */

import { execFile } from "node:child_process";
import { writeFile, unlink } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { promisify } from "node:util";

import { pathToFileURL } from "node:url";
import assert from "node:assert/strict";
const execFileAsync = promisify(execFile);
const PYTHON = process.platform === "win32" ? "python" : "python3";

export function toToolContent(r) {
  const parts = [`exit_code: ${r.exitCode}`];
  if (r.stdout) parts.push(`stdout:\n${r.stdout.trim()}`);
  if (r.stderr) parts.push(`stderr:\n${r.stderr.trim()}`);
  return parts.join("\n");
}

export class SubprocessExecutor {
  constructor(timeoutSeconds = 10) {
    this.timeoutSeconds = timeoutSeconds;
  }

  async run(code, timeoutSeconds) {
    const seconds = timeoutSeconds ?? this.timeoutSeconds;
    const timeout = seconds * 1000;
    const tmpFile = join(tmpdir(), `sandbox_${Date.now()}_${Math.random().toString(36).slice(2)}.py`);
    await writeFile(tmpFile, code, "utf-8");
    try {
      // execFile = no shell = no shell injection
      const { stdout, stderr } = await execFileAsync(PYTHON, [tmpFile], {
        timeout,
        encoding: "utf8",
      });
      return { stdout, stderr, exitCode: 0 };
    } catch (err) {
      if (err.killed || err.signal === "SIGTERM") {
        return {
          stdout: "",
          stderr: `Execution timed out after ${seconds} seconds.`,
          exitCode: 124, // bash timeout(1) convention
        };
      }
      return {
        stdout: err.stdout ?? "",
        stderr: err.stderr ?? String(err),
        exitCode: typeof err.code === "number" ? err.code : 1,
      };
    } finally {
      await unlink(tmpFile).catch(() => {}); // cleanup even on crash
    }
  }
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  const ex = new SubprocessExecutor(3);

  console.log("TEST 1: happy path");
  let r = await ex.run("print(sum(range(101)))");
  console.log(`  ${toToolContent(r)}`);
  assert.ok(r.exitCode === 0 && r.stdout.includes("5050"));

  console.log("\nTEST 2: deliberate NameError (must return, not throw)");
  r = await ex.run("print(undefined_variable)");
  console.log(`  exitCode=${r.exitCode}, stderr contains NameError: ${r.stderr.includes("NameError")}`);
  assert.ok(r.exitCode !== 0 && r.stderr.includes("NameError"));

  console.log("\nTEST 3: infinite loop (must time out at exit code 124, not hang)");
  r = await ex.run("while True: pass");
  console.log(`  exitCode=${r.exitCode}, stderr=${r.stderr.slice(0, 50)}`);
  assert.ok(r.exitCode === 124);

  console.log("\nAll executor checks passed.");
}
